File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import defaultdict, deque
def solution(n, computers):
    answer = 0
    
    graph = defaultdict(list)
    for idx, row in enumerate(computers):
        for j in range(len(row)):
            if idx != j and row[j] == 1:
                graph[idx].append(j)
                
    visited = [False for _ in range(len(computers))]
    
    for i in range(len(visited)):
        if not visited[i]:
            answer += 1
            q = deque()
            q.append(i)
            while q:
                x = q.popleft()
                if visited[x]:
                    continue
                visited[x] = True
                
                for next in graph[x]:
                    q.append(next)
            
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    Queue<Integer> q = new LinkedList<>();
    public int solution(int n, int[][] computers) {
        int answer = 0;
        int[] visited = new int[computers.length];
        for (int i = 0; i < visited.length; i++)
            visited[i] = 0;
        
        for (int i = 0; i < computers.length; i++) {
            if (visited[i] == 0) {
                answer += 1;
                q.add(i);
                while(!q.isEmpty()) {
                    int node = q.poll();
                    if (visited[node] == 1)
                        continue;
                    visited[node] = 1;
                    for (int j = 0; j < computers[node].length; j++) {
                        if (computers[node][j] > 0) {
                            q.add(j);
                        }
                    }
                }
            }
        }
        
        return answer;
    }
}